tsdb/shard.go

```diff
@@ -1302,6 +1306,146 @@ func (a Shards) FieldKeysByPredicate(expr influxql.Expr) (map[string][]string, e
 	return all, nil
 }
 
+// consecutiveAndChildren finds all child nodes of consecutive
+// influxql.BinaryExpr with AND operator nodes ("AND nodes") which are not
+// themselves AND nodes. This may be the root of the tree if the root of the
+// tree is not an AND node.
+type consecutiveAndChildren struct {
+	children []influxql.Node
+}
+
+func (v *consecutiveAndChildren) Visit(node influxql.Node) influxql.Visitor {
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		if n.Op == influxql.AND {
+			return v
+		}
+	case *influxql.ParenExpr:
+		// Parens are essentially a no-op and can be traversed through.
+		return v
+	}
+
+	// If this wasn't a BinaryExpr with an AND operator or a Paren, record this
+	// child node and stop the search for this branch.
+	v.children = append(v.children, node)
+	return nil
+}
+
+// orMeasurementTree determines if a tree (or subtree) represents a grouping of
+// exclusively measurement names OR'd together with EQ operators for the
+// measurements themselves. It collects the list of measurement names
+// encountered and records the validity of the tree.
+type orMeasurementTree struct {
+	measurementKey   string
+	measurementNames []string
+	valid            bool
+}
+
+func (v *orMeasurementTree) Visit(node influxql.Node) influxql.Visitor {
+	// Return early if this tree has already been invalidated - no reason to
+	// continue evaluating at that point.
+	if !v.valid {
+		return nil
+	}
+
+	switch n := node.(type) {
+	case *influxql.BinaryExpr:
+		// A BinaryExpr must have an operation of OR or EQ in a valid tree
+		if n.Op == influxql.OR {
+			// The children of ORs must be either BinaryExprs themselves, or Parens
+			if binaryOrParen(n.LHS) && binaryOrParen(n.RHS) {
```

Comment:
I don't think you need this anymore - you can just `return v` now, if not binary or paren then the visitor will return false (hence invalid) anyway.